### PR TITLE
Add CF ASG tool CLI to work with binding/unbinding security groups to spaces.

### PR DESCRIPTION
**Pull request by apburnes**
Influenced from research finding in https://github.com/cloud-gov/cg-deploy-cf/issues/565
 
## Changes proposed in this pull request:
- Add cli tool to work with binding all spaces to ASGs across the platform
- [x] Bind a security group to all spaces
- [x] Unbind a security group to all spaces
- [X] Check all spaces security groups  

## security considerations
none

### CODE REVIEW

File: asg-tool/__main__.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="ASG tool to bind and check ASG's applied to CF spaces."
+    )
+
+    subparsers = parser.add_subparsers(
+        dest="command", help="Run <command> help for more information."
+    )
+
+    subparsers.add_parser("spaces-asg", help="List all spaces and their ASG's.")
```
Comment (bengerman13) [inline asg-tool/__main__.py:15]: Meta: would it make sense to include an unbind command here? That way we could quickly roll back if we discover a problem?

Comment (apburnes) [reply asg-tool/__main__.py:15]: Totally. Bind, unbind and audit/check are three commands I'd imagine we'd need.

**Review by ccostino — COMMENTED** [review]
Thanks, @apburnes!  This looks really nice and is very clean, :high-five: sir!

I had a couple of general questions just to double check my understanding of things more than anything.

File: asg-tool/asg_tool.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    results = cf_curl_results_handler(output)
+    pagination = results["pagination"]
+    next_page = pagination["next"]
+    resources = results["resources"]
+    joined_resources = acc_resources + resources
+
+    if next_page:
+        parsed_api = urlparse(next_page["href"])
+        next_path = f"{parsed_api.path}?{parsed_api.query}"
+        return cf_curl_get(
```
[the remark this answers is not in the document]
Comment (apburnes) [reply asg-tool/asg_tool.py:67]: I think it's just my javascript experience influencing function design.

File: asg-tool/asg_tool.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    return filtered
+
+
+def get_asg_guid(asg_name: str) -> str:
+    results = cf_curl_get(f"/v3/security_groups?names={asg_name}")
+    pick_asg = list(filter(lambda asg: asg["name"] == asg_name, results))
+    filtered = filter_resources(pick_asg)
+
+    if len(filtered) < 1:
+        return None
```
[the remark this answers is not in the document]
Comment (apburnes) [reply asg-tool/asg_tool.py:107]: I'm in the same boat with typing. I'll poke around and see if I can find something.

[the remark this answers is not in the document]
Comment (apburnes) [reply asg-tool/asg_tool.py:107]: I think it's this https://docs.python.org/3/library/typing.html#typing.Union

File: asg-tool/tests.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        data = {"test": "data"}
+        stdout = {"data": {"message": "Success"}}
+        mock_call.return_value = SubprocessResult(stdout=stdout)
+        result = asg_tool.cf_curl_post("/v3/test/1", data)
+        self.assertEqual(result, stdout["data"])
+
+
+class TestCheckSpaceASG(unittest.TestCase):
+    def test_curl_space_asg(self):
+        pass
```
[the remark this answers is not in the document]
Comment (apburnes) [reply asg-tool/tests.py:177]: Totally forgot this one.

File: asg-tool/asg_tool.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ccostino) [reply asg-tool/asg_tool.py:107]: Ah I see, that or [`typing.Optional`](https://docs.python.org/3/library/typing.html#typing.Optional)

File: asg-tool/asg_tool.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    return updated_item
+
+
+def filter_resources(resources: list) -> list:
+    filtered = []
+
+    for item in resources:
+        filtered_item = filter_resource_item(item)
+
+        if filtered_item == None:
```
Comment (bengerman13) [inline asg-tool/asg_tool.py:25]: this should be `is None`

Comment (apburnes) [reply asg-tool/asg_tool.py:25]: 👍

File: asg-tool/asg_tool.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (apburnes) [reply asg-tool/asg_tool.py:108]: Added `typing.Optional`

File: asg-tool/tests.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (apburnes) [reply asg-tool/tests.py]: Added tests

**Review by bengerman13 — APPROVED** [review]

**Pull request merged**